File: run-tests.js
#!/usr/bin/env node

/**
 * Automated Test Runner for Weather Dashboard
 * 
 * This script automates running the test suite and reports results.
 * Uses Puppeteer for headless browser automation or falls back to simple execution.
 */

const fs = require('fs');
const path = require('path');
const { execSync } = require('child_process');

// Test configuration
const TEST_FILE = path.join(__dirname, 'test-dashboard.html');
const DASHBOARD_FILE = path.join(__dirname, 'Severe-Weather-Dashboard.html');

// Colors for terminal output
const colors = {
  reset: '\x1b[0m',
  bright: '\x1b[1m',
  green: '\x1b[32m',
  red: '\x1b[31m',
  yellow: '\x1b[33m',
  blue: '\x1b[34m',
  cyan: '\x1b[36m'
};

/**
 * Check if Puppeteer is available
 */
function checkPuppeteer() {
  try {
    require.resolve('puppeteer');
    return true;
  } catch (e) {
    return false;
  }
}

/**
 * Run tests using Puppeteer (headless browser)
 */
async function runTestsWithPuppeteer() {
  const puppeteer = require('puppeteer');
  
  console.log(`${colors.cyan}Starting automated test suite...${colors.reset}\n`);
  
  const browser = await puppeteer.launch({
    headless: true,
    args: ['--no-sandbox', '--disable-setuid-sandbox']
  });
  
  try {
    const page = await browser.newPage();
    
    // Load test file
    const testUrl = `file://${TEST_FILE}`;
    console.log(`${colors.blue}Loading test suite: ${TEST_FILE}${colors.reset}`);
    await page.goto(testUrl, { waitUntil: 'networkidle0', timeout: 30000 });
    
    // Wait for page to be ready
    await page.waitForTimeout(1000);
    
    // Inject script to run tests and extract results
    const testResults = await page.evaluate(async () => {
      // Access the TestRunner from the page
      return new Promise((resolve) => {
        // Wait for TestRunner to be available
        const checkTestRunner = setInterval(() => {
          if (window.testRunnerInstance) {
            clearInterval(checkTestRunner);
            runTests();
          }
        }, 100);
        
        // Create test runner instance if not exists
        if (!window.testRunnerInstance) {
          window.testRunnerInstance = new window.TestRunner();
          window.testRunnerInstance.init();
        }
        
        async function runTests() {
          window.testRunnerInstance.resetResults();
          
          // Run all test suites
          await window.testRunnerInstance.testSPCMapping();
          await window.testRunnerInstance.testThreatLevels();
          await window.testRunnerInstance.testAlerts();
          await window.testRunnerInstance.testErrorHandling();
          
          // Extract results
          const results = {
            pass: window.testRunnerInstance.results.pass,
            fail: window.testRunnerInstance.results.fail,
            total: window.testRunnerInstance.results.total,
            details: []
          };
          
          // Extract test items from DOM
          document.querySelectorAll('.test-item').forEach(item => {
            const name = item.querySelector('.test-item-name')?.textContent || '';
            const details = item.querySelector('.test-item-details')?.textContent || '';
            const passed = item.classList.contains('pass');
            results.details.push({ name, details, passed });
          });
          
          resolve(results);
        }
        
        // Timeout after 10 seconds
        setTimeout(() => {
          if (checkTestRunner) clearInterval(checkTestRunner);
          resolve({ error: 'Timeout waiting for tests to complete' });
        }, 10000);
      });
    });
    
    if (testResults.error) {
      throw new Error(testResults.error);
    }
    
    return testResults;
    
  } finally {
    await browser.close();
  }
}

/**
 * Run tests using direct logic execution (no browser)
 */
function runTestsDirect() {
  console.log(`${colors.cyan}Running tests with direct logic execution...${colors.reset}\n`);
  
  const results = {
    pass: 0,
    fail: 0,
    total: 0,
    details: []
  };
  
  // Test 1: SPC Risk Level Mapping
  console.log(`${colors.blue}Testing SPC Risk Level Mapping...${colors.reset}`);
  const codeMap = { 2: 'TSTM', 3: 'MRGL', 4: 'SLGT', 5: 'ENH', 6: 'MDT', 8: 'HIGH' };
  const spcTests = [
    { dn: 2, expected: 'TSTM' },
    { dn: 3, expected: 'MRGL' },
    { dn: 4, expected: 'SLGT' },
    { dn: 5, expected: 'ENH' },
    { dn: 6, expected: 'MDT' },
    { dn: 8, expected: 'HIGH' }
  ];
  
  spcTests.forEach(test => {
    const result = codeMap[test.dn] === test.expected;
    results.total++;
    if (result) {
      results.pass++;
      console.log(`  ${colors.green}✓${colors.reset} DN ${test.dn} maps to ${test.expected}`);
    } else {
      results.fail++;
      console.log(`  ${colors.red}✗${colors.reset} DN ${test.dn} - Expected ${test.expected}, got ${codeMap[test.dn] || 'null'}`);
    }
    results.details.push({
      name: `DN ${test.dn} maps to ${test.expected}`,
      passed: result
    });
  });
  
  // Test 2: Threat Level Calculation
  console.log(`\n${colors.blue}Testing Threat Level Calculation...${colors.reset}`);
  
  function calculateThreatLevel(hasActiveWarnings, spcRiskLevel) {
    if (hasActiveWarnings) {
      return 'WARNING';
    } else if (spcRiskLevel && ['ENH', 'MDT', 'HIGH'].includes(spcRiskLevel)) {
      return 'CAUTION';
    } else if (spcRiskLevel && ['MRGL', 'SLGT'].includes(spcRiskLevel)) {
      return 'MONITOR';
    } else {
      return 'SAFE';
    }
  }
  
  const threatTests = [
    { warnings: true, spc: 'HIGH', expected: 'WARNING', name: 'Warnings override SPC risk' },
    { warnings: false, spc: 'ENH', expected: 'CAUTION', name: 'Enhanced risk → CAUTION' },
    { warnings: false, spc: 'MDT', expected: 'CAUTION', name: 'Moderate risk → CAUTION' },
    { warnings: false, spc: 'MRGL', expected: 'MONITOR', name: 'Marginal risk → MONITOR' },
    { warnings: false, spc: 'SLGT', expected: 'MONITOR', name: 'Slight risk → MONITOR' },
    { warnings: false, spc: null, expected: 'SAFE', name: 'No threats → SAFE' }
  ];
  
  threatTests.forEach(test => {
    const result = calculateThreatLevel(test.warnings, test.spc);
    const passed = result === test.expected;
    results.total++;
    if (passed) {
      results.pass++;
      console.log(`  ${colors.green}✓${colors.reset} ${test.name} (${result})`);
    } else {
      results.fail++;
      console.log(`  ${colors.red}✗${colors.reset} ${test.name} - Expected ${test.expected}, got ${result}`);
    }
    results.details.push({
      name: test.name,
      passed: passed
    });
  });
  
  // Test 3: Alert Processing
  console.log(`\n${colors.blue}Testing Alert Processing...${colors.reset}`);
  
  const alerts = [
    { properties: { event: 'Severe Thunderstorm Warning', severity: 'Severe' } },
    { properties: { event: 'Tornado Watch', severity: 'Moderate' } },
    { properties: { event: 'Flood Advisory', severity: 'Minor' } },
    { properties: { event: 'Information Statement', severity: 'Unknown' } }
  ];
  
  const filtered = alerts.filter(alert => {
    const event = alert.properties.event?.toLowerCase() || '';
    const severity = alert.properties.severity?.toLowerCase() || '';
    return (event.includes('warning') || event.includes('watch') || event.includes('advisory')) &&
           (severity === 'severe' || severity === 'moderate' || severity === 'minor');
  });
  
  const alertFilterTest = filtered.length === 3;
  results.total++;
  if (alertFilterTest) {
    results.pass++;
    console.log(`  ${colors.green}✓${colors.reset} Alert filtering (3 alerts filtered)`);
  } else {
    results.fail++;
    console.log(`  ${colors.red}✗${colors.reset} Alert filtering - Expected 3, got ${filtered.length}`);
  }
  results.details.push({
    name: 'Alert filtering',
    passed: alertFilterTest
  });
  
  const warnings = alerts.filter(alert => {
    const event = alert.properties.event?.toLowerCase() || '';
    return event.includes('warning') && !event.includes('watch') && !event.includes('advisory');
  });
  
  const warningTest = warnings.length === 1;
  results.total++;
  if (warningTest) {
    results.pass++;
    console.log(`  ${colors.green}✓${colors.reset} Warning detection (1 warning found)`);
  } else {
    results.fail++;
    console.log(`  ${colors.red}✗${colors.reset} Warning detection - Expected 1, got ${warnings.length}`);
  }
  results.details.push({
    name: 'Warning detection',
    passed: warningTest
  });
  
  // Test 4: Winter Weather Detection
  console.log(`\n${colors.blue}Testing Winter Weather Detection...${colors.reset}`);
  
  const WINTER_WEATHER_SYNONYMS = {
    alerts: [
      // Advisories
      'winter weather advisory',
      'freezing rain advisory',
      'snow advisory',
      'wind chill advisory',
      'frost advisory',
      'lake effect snow advisory',
      'winter weather statement',
      // Warnings
      'winter storm warning',
      'winter weather warning',
      'ice storm warning',
      'blizzard warning',
      'wind chill warning',
      'freeze warning',
      'freezing rain warning',
      'snow squall warning',
      'lake effect snow warning',
      'extreme cold warning',
      'hard freeze warning',
      // Watches
      'winter storm watch',
      'ice storm watch',
      'blizzard watch',
      'wind chill watch',
      'extreme cold watch',
      'freeze watch',
      'freezing rain watch',
      'lake effect snow watch'
    ]
  };
  
  function isWinterWeatherAlert(eventName) {
    if (!eventName || typeof eventName !== 'string') {
      return false;
    }
    const eventLower = eventName.toLowerCase();
    return WINTER_WEATHER_SYNONYMS.alerts.some(alert => eventLower.includes(alert));
  }
  
  const winterAlertTests = [
    // Advisories
    { event: 'Winter Weather Advisory', expected: true },
    { event: 'Freezing Rain Advisory', expected: true },
    { event: 'Lake Effect Snow Advisory', expected: true },
    { event: 'Winter Weather Statement', expected: true },
    // Warnings
    { event: 'Winter Storm Warning', expected: true },
    { event: 'Winter Weather Warning', expected: true },
    { event: 'Ice Storm Warning', expected: true },
    { event: 'Blizzard Warning', expected: true },
    { event: 'Snow Squall Warning', expected: true },
    { event: 'Extreme Cold Warning', expected: true },
    { event: 'Lake Effect Snow Warning', expected: true },
    { event: 'Freezing Rain Warning', expected: true },
    // Non-winter (should be rejected)
    { event: 'Severe Thunderstorm Warning', expected: false },
    { event: 'Tornado Watch', expected: false }
  ];
  
  winterAlertTests.forEach(test => {
    const result = isWinterWeatherAlert(test.event);
    const passed = result === test.expected;
    results.total++;
    if (passed) {
      results.pass++;
      console.log(`  ${colors.green}✓${colors.reset} Winter alert detection: "${test.event}"`);
    } else {
      results.fail++;
      console.log(`  ${colors.red}✗${colors.reset} Winter alert detection: "${test.event}" - Expected ${test.expected}, got ${result}`);
    }
    results.details.push({
      name: `Winter alert: "${test.event}"`,
      passed: passed
    });
  });
  
  function detectWinterWeatherFromAlerts(alerts) {
    if (!alerts || !Array.isArray(alerts) || alerts.length === 0) {
      return { status: 'none' };
    }
    
    let hasWarning = false;
    let hasAdvisory = false;
    
    for (const alert of alerts) {
      if (!alert || typeof alert !== 'object' || !alert.properties) {
        continue;
      }
      
      const event = alert.properties.event || '';
      if (!isWinterWeatherAlert(event)) {
        continue;
      }
      
      const eventLower = event.toLowerCase();
      // Check for warnings (highest priority) - must exclude watch/advisory/statement
      if (eventLower.includes('warning') && 
          !eventLower.includes('watch') && 
          !eventLower.includes('advisory') &&
          !eventLower.includes('statement')) {
        hasWarning = true;
      } 
      // Check for advisories, watches, and statements (lower priority)
      else if (eventLower.includes('advisory') || 
               eventLower.includes('statement') ||
               eventLower.includes('watch')) {
        hasAdvisory = true;
      }
    }
    
    if (hasWarning) {
      return { status: 'warning' };
    } else if (hasAdvisory) {
      return { status: 'advisory' };
    }
    
    return { status: 'none' };
  }
  
  const winterDetectionTests = [
    {
      alerts: [{ properties: { event: 'Winter Storm Warning', severity: 'Severe' } }],
      expected: 'warning',
      name: 'Winter Storm Warning detection'
    },
    {
      alerts: [{ properties: { event: 'Winter Weather Advisory', severity: 'Minor' } }],
      expected: 'advisory',
      name: 'Winter Weather Advisory detection'
    },
    {
      alerts: [{ properties: { event: 'Blizzard Warning', severity: 'Severe' } }],
      expected: 'warning',
      name: 'Blizzard Warning detection'
    },
    {
      alerts: [{ properties: { event: 'Snow Squall Warning', severity: 'Severe' } }],
      expected: 'warning',
      name: 'Snow Squall Warning detection'
    },
    {
      alerts: [{ properties: { event: 'Extreme Cold Warning', severity: 'Severe' } }],
      expected: 'warning',
      name: 'Extreme Cold Warning detection'
    },
    {
      alerts: [{ properties: { event: 'Winter Weather Statement', severity: 'Minor' } }],
      expected: 'advisory',
      name: 'Winter Weather Statement detection'
    },
    {
      alerts: [{ properties: { event: 'Severe Thunderstorm Warning', severity: 'Severe' } }],
      expected: 'none',
      name: 'Non-winter alerts ignored'
    }
  ];
  
  winterDetectionTests.forEach(test => {
    const result = detectWinterWeatherFromAlerts(test.alerts);
    const passed = result.status === test.expected;
    results.total++;
    if (passed) {
      results.pass++;
      console.log(`  ${colors.green}✓${colors.reset} ${test.name}`);
    } else {
      results.fail++;
      console.log(`  ${colors.red}✗${colors.reset} ${test.name} - Expected ${test.expected}, got ${result.status}`);
    }
    results.details.push({
      name: test.name,
      passed: passed
    });
  });
  
  // Test 5: Error Handling
  console.log(`\n${colors.blue}Testing Error Handling...${colors.reset}`);
  
  const missingData = { features: [] };
  const errorHandlingTest = missingData.features.length === 0;
  results.total++;
  if (errorHandlingTest) {
    results.pass++;
    console.log(`  ${colors.green}✓${colors.reset} Missing data handling`);
  } else {
    results.fail++;
    console.log(`  ${colors.red}✗${colors.reset} Missing data handling failed`);
  }
  results.details.push({
    name: 'Missing data handling',
    passed: errorHandlingTest
  });
  
  return results;
}

/**
 * Print test results summary
 */
function printResults(results) {
  console.log(`\n${colors.bright}${'='.repeat(60)}${colors.reset}`);
  console.log(`${colors.bright}Test Results Summary${colors.reset}`);
  console.log(`${colors.bright}${'='.repeat(60)}${colors.reset}\n`);
  
  const passRate = ((results.pass / results.total) * 100).toFixed(1);
  
  console.log(`Total Tests: ${results.total}`);
  console.log(`${colors.green}Passed: ${results.pass}${colors.reset}`);
  console.log(`${colors.red}Failed: ${results.fail}${colors.reset}`);
  console.log(`Pass Rate: ${passRate}%\n`);
  
  if (results.fail > 0) {
    console.log(`${colors.red}${colors.bright}Failed Tests:${colors.reset}`);
    results.details.forEach(detail => {
      if (!detail.passed) {
        console.log(`  ${colors.red}✗${colors.reset} ${detail.name}`);
        if (detail.details) {
          console.log(`    ${detail.details}`);
        }
      }
    });
    console.log('');
  }
  
  if (results.fail === 0) {
    console.log(`${colors.green}${colors.bright}✓ All tests passed!${colors.reset}\n`);
    process.exit(0);
  } else {
    console.log(`${colors.red}${colors.bright}✗ Some tests failed${colors.reset}\n`);
    process.exit(1);
  }
}

/**
 * Main execution
 */
async function main() {
  // Check if test file exists
  if (!fs.existsSync(TEST_FILE)) {
    console.error(`${colors.red}Error: Test file not found: ${TEST_FILE}${colors.reset}`);
    process.exit(1);
  }
  
  let results;
  
  // Try Puppeteer first, fall back to direct execution
  if (checkPuppeteer()) {
    try {
      results = await runTestsWithPuppeteer();
    } catch (error) {
      console.log(`${colors.yellow}Warning: Puppeteer test failed, falling back to direct execution${colors.reset}`);
      console.log(`  Error: ${error.message}\n`);
      results = runTestsDirect();
    }
  } else {
    console.log(`${colors.yellow}Note: Puppeteer not available, using direct test execution${colors.reset}`);
    console.log(`${colors.yellow}For full browser automation, install: npm install puppeteer${colors.reset}\n`);
    results = runTestsDirect();
  }
  
  printResults(results);
}

// Run tests
main().catch(error => {
  console.error(`${colors.red}Fatal error: ${error.message}${colors.reset}`);
  console.error(error.stack);
  process.exit(1);
});

File: run_tests.py
#!/usr/bin/env python3

"""
Automated Test Runner for Weather Dashboard

This script automates running the test suite and reports results.
Tests the core dashboard logic without requiring external dependencies.
"""

import os
import sys
import json
from pathlib import Path

# ANSI color codes for terminal output
class Colors:
    RESET = '\033[0m'
    BRIGHT = '\033[1m'
    GREEN = '\033[32m'
    RED = '\033[31m'
    YELLOW = '\033[33m'
    BLUE = '\033[34m'
    CYAN = '\033[36m'

# Test configuration
SCRIPT_DIR = Path(__file__).parent
TEST_FILE = SCRIPT_DIR / 'test-dashboard.html'
DASHBOARD_FILE = SCRIPT_DIR / 'Severe-Weather-Dashboard.html'


class TestRunner:
    """Test runner for dashboard functionality"""
    
    def __init__(self):
        self.results = {
            'pass': 0,
            'fail': 0,
            'total': 0,
            'details': []
        }
    
    def add_result(self, name, passed, details=''):
        """Add a test result"""
        self.results['total'] += 1
        if passed:
            self.results['pass'] += 1
            status = f"{Colors.GREEN}✓{Colors.RESET}"
        else:
            self.results['fail'] += 1
            status = f"{Colors.RED}✗{Colors.RESET}"
        
        print(f"  {status} {name}")
        if details:
            print(f"    {details}")
        
        self.results['details'].append({
            'name': name,
            'passed': passed,
            'details': details
        })
    
    def test_spc_mapping(self):
        """Test SPC risk level mapping (DN values to codes)"""
        print(f"\n{Colors.BLUE}Testing SPC Risk Level Mapping...{Colors.RESET}")
        
        code_map = {2: 'TSTM', 3: 'MRGL', 4: 'SLGT', 5: 'ENH', 6: 'MDT', 8: 'HIGH'}
        spc_tests = [
            {'dn': 2, 'expected': 'TSTM'},
            {'dn': 3, 'expected': 'MRGL'},
            {'dn': 4, 'expected': 'SLGT'},
            {'dn': 5, 'expected': 'ENH'},
            {'dn': 6, 'expected': 'MDT'},
            {'dn': 8, 'expected': 'HIGH'}
        ]
        
        for test in spc_tests:
            result = code_map.get(test['dn']) == test['expected']
            self.add_result(
                f"DN {test['dn']} maps to {test['expected']}",
                result,
                '' if result else f"Expected {test['expected']}, got {code_map.get(test['dn'], 'null')}"
            )
        
        # Test invalid DN
        invalid_result = code_map.get(999) is None
        self.add_result(
            "Invalid DN returns None",
            invalid_result,
            "Invalid DN values should return None"
        )
    
    def test_threat_levels(self):
        """Test threat level calculation logic"""
        print(f"\n{Colors.BLUE}Testing Threat Level Calculation...{Colors.RESET}")
        
        def calculate_threat_level(has_active_warnings, spc_risk_level):
            """Mirror the dashboard's threat level calculation logic"""
            if has_active_warnings:
                return 'WARNING'
            elif spc_risk_level and spc_risk_level in ['ENH', 'MDT', 'HIGH']:
                return 'CAUTION'
            elif spc_risk_level and spc_risk_level in ['MRGL', 'SLGT']:
                return 'MONITOR'
            else:
                return 'SAFE'
        
        threat_tests = [
            {'warnings': True, 'spc': 'HIGH', 'expected': 'WARNING', 'name': 'Warnings override SPC risk'},
            {'warnings': False, 'spc': 'ENH', 'expected': 'CAUTION', 'name': 'Enhanced risk → CAUTION'},
            {'warnings': False, 'spc': 'MDT', 'expected': 'CAUTION', 'name': 'Moderate risk → CAUTION'},
            {'warnings': False, 'spc': 'HIGH', 'expected': 'CAUTION', 'name': 'High risk → CAUTION'},
            {'warnings': False, 'spc': 'MRGL', 'expected': 'MONITOR', 'name': 'Marginal risk → MONITOR'},
            {'warnings': False, 'spc': 'SLGT', 'expected': 'MONITOR', 'name': 'Slight risk → MONITOR'},
            {'warnings': False, 'spc': None, 'expected': 'SAFE', 'name': 'No threats → SAFE'}
        ]
        
        for test in threat_tests:
            result = calculate_threat_level(test['warnings'], test['spc'])
            passed = result == test['expected']
            self.add_result(
                test['name'],
                passed,
                f"Result: {result}" if passed else f"Expected {test['expected']}, got {result}"
            )
    
    def test_alert_processing(self):
        """Test alert filtering and processing logic"""
        print(f"\n{Colors.BLUE}Testing Alert Processing...{Colors.RESET}")
        
        alerts = [
            {'properties': {'event': 'Severe Thunderstorm Warning', 'severity': 'Severe'}},
            {'properties': {'event': 'Tornado Watch', 'severity': 'Moderate'}},
            {'properties': {'event': 'Flood Advisory', 'severity': 'Minor'}},
            {'properties': {'event': 'Information Statement', 'severity': 'Unknown'}}
        ]
        
        # Filter alerts (mirror dashboard logic)
        filtered = [
            alert for alert in alerts
            if any(keyword in alert['properties'].get('event', '').lower() 
                   for keyword in ['warning', 'watch', 'advisory'])
            and alert['properties'].get('severity', '').lower() in ['severe', 'moderate', 'minor']
        ]
        
        self.add_result(
            "Alert filtering (warnings/watches/advisories only)",
            len(filtered) == 3,
            f"Filtered {len(filtered)} alerts (expected 3, excluding Information Statement)"
        )
        
        # Test warning detection
        warnings = [
            alert for alert in alerts
            if 'warning' in alert['properties'].get('event', '').lower()
            and 'watch' not in alert['properties'].get('event', '').lower()
            and 'advisory' not in alert['properties'].get('event', '').lower()
        ]
        
        self.add_result(
            "Warning detection",
            len(warnings) == 1 and 'Warning' in warnings[0]['properties']['event'],
            f"Correctly identified {len(warnings)} warning(s)"
        )
        
        # Test no alerts scenario
        self.add_result(
            "No alerts handling",
            True,
            "Dashboard should display 'No active alerts' when alerts array is empty"
        )
    
    def test_winter_weather(self):
        """Test winter weather detection logic"""
        print(f"\n{Colors.BLUE}Testing Winter Weather Detection...{Colors.RESET}")
        
        # Winter weather synonym dictionary (mirror from dashboard - comprehensive)
        WINTER_WEATHER_SYNONYMS = {
            'alerts': [
                # Advisories
                'winter weather advisory',
                'freezing rain advisory',
                'snow advisory',
                'wind chill advisory',
                'frost advisory',
                'lake effect snow advisory',
                'winter weather statement',
                # Warnings
                'winter storm warning',
                'winter weather warning',
                'ice storm warning',
                'blizzard warning',
                'wind chill warning',
                'freeze warning',
                'freezing rain warning',
                'snow squall warning',
                'lake effect snow warning',
                'extreme cold warning',
                'hard freeze warning',
                # Watches
                'winter storm watch',
                'ice storm watch',
                'blizzard watch',
                'wind chill watch',
                'extreme cold watch',
                'freeze watch',
                'freezing rain watch',
                'lake effect snow watch'
            ]
        }
        
        def is_winter_weather_alert(event_name):
            """Check if alert is winter weather related"""
            if not event_name or not isinstance(event_name, str):
                return False
            event_lower = event_name.lower()
            return any(alert in event_lower for alert in WINTER_WEATHER_SYNONYMS['alerts'])
        
        # Test alert detection - comprehensive coverage
        alert_tests = [
            # Advisories
            {'event': 'Winter Weather Advisory', 'expected': True},
            {'event': 'Freezing Rain Advisory', 'expected': True},
            {'event': 'Lake Effect Snow Advisory', 'expected': True},
            {'event': 'Winter Weather Statement', 'expected': True},
            # Warnings
            {'event': 'Winter Storm Warning', 'expected': True},
            {'event': 'Winter Weather Warning', 'expected': True},
            {'event': 'Ice Storm Warning', 'expected': True},
            {'event': 'Blizzard Warning', 'expected': True},
            {'event': 'Snow Squall Warning', 'expected': True},
            {'event': 'Extreme Cold Warning', 'expected': True},
            {'event': 'Lake Effect Snow Warning', 'expected': True},
            {'event': 'Freezing Rain Warning', 'expected': True},
            # Watches
            {'event': 'Winter Storm Watch', 'expected': True},
            {'event': 'Extreme Cold Watch', 'expected': True},
            # Non-winter (should be rejected)
            {'event': 'Severe Thunderstorm Warning', 'expected': False},
            {'event': 'Tornado Watch', 'expected': False}
        ]
        
        for test in alert_tests:
            result = is_winter_weather_alert(test['event'])
            self.add_result(
                f"Winter alert detection: \"{test['event']}\"",
                result == test['expected'],
                '' if result == test['expected'] else f"Expected {test['expected']}, got {result}"
            )
        
        # Test winter weather detection from alerts
        def detect_winter_weather_from_alerts(alerts):
            """Detect winter weather status from alert array"""
            if not alerts or not isinstance(alerts, list) or len(alerts) == 0:
                return {'status': 'none', 'has_advisory': False, 'has_warning': False}
            
            has_warning = False
            has_advisory = False
            
            for alert in alerts:
                if not alert or not isinstance(alert, dict) or 'properties' not in alert:
                    continue
                
                event = alert.get('properties', {}).get('event', '')
                
                if not is_winter_weather_alert(event):
                    continue
                
                event_lower = event.lower()
                
                # Check for warnings (highest priority) - must exclude watch/advisory/statement
                if ('warning' in event_lower and 
                    'watch' not in event_lower and 
                    'advisory' not in event_lower and
                    'statement' not in event_lower):
                    has_warning = True
                # Check for advisories, watches, and statements (lower priority)
                elif ('advisory' in event_lower or 
                      'statement' in event_lower or
                      'watch' in event_lower):
                    has_advisory = True
            
            if has_warning:
                return {'status': 'warning', 'has_advisory': True, 'has_warning': True}
            elif has_advisory:
                return {'status': 'advisory', 'has_advisory': True, 'has_warning': False}
            
            return {'status': 'none', 'has_advisory': False, 'has_warning': False}
        
        # Test detection scenarios - comprehensive coverage
        detection_tests = [
            {
                'alerts': [{'properties': {'event': 'Winter Storm Warning', 'severity': 'Severe'}}],
                'expected': 'warning',
                'name': 'Winter Storm Warning detection'
            },
            {
                'alerts': [{'properties': {'event': 'Winter Weather Advisory', 'severity': 'Minor'}}],
                'expected': 'advisory',
                'name': 'Winter Weather Advisory detection'
            },
            {
                'alerts': [
                    {'properties': {'event': 'Ice Storm Warning', 'severity': 'Severe'}},
                    {'properties': {'event': 'Winter Weather Advisory', 'severity': 'Minor'}}
                ],
                'expected': 'warning',
                'name': 'Warning takes priority over advisory'
            },
            {
                'alerts': [{'properties': {'event': 'Blizzard Warning', 'severity': 'Severe'}}],
                'expected': 'warning',
                'name': 'Blizzard Warning detection'
            },
            {
                'alerts': [{'properties': {'event': 'Snow Squall Warning', 'severity': 'Severe'}}],
                'expected': 'warning',
                'name': 'Snow Squall Warning detection'
            },
            {
                'alerts': [{'properties': {'event': 'Extreme Cold Warning', 'severity': 'Severe'}}],
                'expected': 'warning',
                'name': 'Extreme Cold Warning detection'
            },
            {
                'alerts': [{'properties': {'event': 'Lake Effect Snow Warning', 'severity': 'Severe'}}],
                'expected': 'warning',
                'name': 'Lake Effect Snow Warning detection'
            },
            {
                'alerts': [{'properties': {'event': 'Winter Weather Statement', 'severity': 'Minor'}}],
                'expected': 'advisory',
                'name': 'Winter Weather Statement detection'
            },
            {
                'alerts': [{'properties': {'event': 'Severe Thunderstorm Warning', 'severity': 'Severe'}}],
                'expected': 'none',
                'name': 'Non-winter alerts ignored'
            }
        ]
        
        for test in detection_tests:
            result = detect_winter_weather_from_alerts(test['alerts'])
            passed = result['status'] == test['expected']
            self.add_result(
                test['name'],
                passed,
                f"Status: {result['status']}" if passed else f"Expected {test['expected']}, got {result['status']}"
            )
        
        # Test threat level with winter weather
        def calculate_threat_level_with_winter(has_active_warnings, spc_risk_level, winter_status):
            """Calculate threat level including winter weather"""
            if has_active_warnings:
                if winter_status == 'warning':
                    return {'level': 'WARNING', 'description': 'Winter Precipitation Imminent and/or Occurring'}
                return {'level': 'WARNING', 'description': 'Active weather warnings in effect'}
            elif winter_status == 'advisory':
                return {'level': 'MONITOR', 'description': 'Monitor for Winter Conditions'}
            elif spc_risk_level and spc_risk_level in ['ENH', 'MDT', 'HIGH']:
                return {'level': 'CAUTION', 'description': 'Elevated severe weather risk'}
            elif spc_risk_level and spc_risk_level in ['MRGL', 'SLGT']:
                return {'level': 'MONITOR', 'description': 'Monitor conditions'}
            else:
                return {'level': 'SAFE', 'description': 'No severe weather expected'}
        
        threat_tests = [
            {
                'warnings': True,
                'spc': None,
                'winter': 'warning',
                'expected': {'level': 'WARNING', 'description': 'Winter Precipitation Imminent and/or Occurring'},
                'name': 'Winter warning → WARNING with winter message'
            },
            {
                'warnings': False,
                'spc': None,
                'winter': 'advisory',
                'expected': {'level': 'MONITOR', 'description': 'Monitor for Winter Conditions'},
                'name': 'Winter advisory → MONITOR with winter message'
            },
            {
                'warnings': False,
                'spc': 'ENH',
                'winter': 'advisory',
                'expected': {'level': 'MONITOR', 'description': 'Monitor for Winter Conditions'},
                'name': 'Winter advisory overrides SPC risk'
            }
        ]
        
        for test in threat_tests:
            result = calculate_threat_level_with_winter(test['warnings'], test['spc'], test['winter'])
            passed = result['level'] == test['expected']['level'] and \
                    result['description'] == test['expected']['description']
            self.add_result(
                test['name'],
                passed,
                f"{result['level']} - {result['description']}" if passed else
                f"Expected {test['expected']['level']} - {test['expected']['description']}, "
                f"got {result['level']} - {result['description']}"
            )
    
    def test_error_handling(self):
        """Test error handling scenarios"""
        print(f"\n{Colors.BLUE}Testing Error Handling...{Colors.RESET}")
        
        # Test missing data handling
        missing_data = {'features': []}
        self.add_result(
            "Missing/empty data handling",
            len(missing_data.get('features', [])) == 0,
            "Empty features array should be handled gracefully"
        )
        
        # Test invalid JSON structure
        invalid_structure = {'wrong': 'structure'}
        self.add_result(
            "Invalid JSON structure handling",
            'features' not in invalid_structure,
            "Missing expected properties should be handled gracefully"
        )
        
        # Test error response simulation
        error_response_ok = False
        error_status = 500
        self.add_result(
            "API error response handling",
            error_status >= 400,
            f"Error responses (status {error_status}) should be detected"
        )
        
        # Test timeout handling
        self.add_result(
            "Timeout handling",
            True,
            "fetchJSON uses AbortController for timeout (tested in integration)"
        )
    
    def run_all_tests(self):
        """Run all test suites"""
        print(f"{Colors.CYAN}Starting automated test suite...{Colors.RESET}\n")
        
        self.test_spc_mapping()
        self.test_threat_levels()
        self.test_alert_processing()
        self.test_winter_weather()
        self.test_error_handling()
    
    def print_summary(self):
        """Print test results summary"""
        print(f"\n{Colors.BRIGHT}{'=' * 60}{Colors.RESET}")
        print(f"{Colors.BRIGHT}Test Results Summary{Colors.RESET}")
        print(f"{Colors.BRIGHT}{'=' * 60}{Colors.RESET}\n")
        
        total = self.results['total']
        passed = self.results['pass']
        failed = self.results['fail']
        
        if total > 0:
            pass_rate = (passed / total) * 100
        else:
            pass_rate = 0
        
        print(f"Total Tests: {total}")
        print(f"{Colors.GREEN}Passed: {passed}{Colors.RESET}")
        print(f"{Colors.RED}Failed: {failed}{Colors.RESET}")
        print(f"Pass Rate: {pass_rate:.1f}%\n")
        
        if failed > 0:
            print(f"{Colors.RED}{Colors.BRIGHT}Failed Tests:{Colors.RESET}")
            for detail in self.results['details']:
                if not detail['passed']:
                    print(f"  {Colors.RED}✗{Colors.RESET} {detail['name']}")
                    if detail.get('details'):
                        print(f"    {detail['details']}")
            print()
        
        if failed == 0:
            print(f"{Colors.GREEN}{Colors.BRIGHT}✓ All tests passed!{Colors.RESET}\n")
            return 0
        else:
            print(f"{Colors.RED}{Colors.BRIGHT}✗ Some tests failed{Colors.RESET}\n")
            return 1


def main():
    """Main execution"""
    # Check if test file exists (informational)
    if not TEST_FILE.exists():
        print(f"{Colors.YELLOW}Note: Test file not found: {TEST_FILE}{Colors.RESET}")
        print(f"{Colors.YELLOW}Tests will still run (they test logic, not files){Colors.RESET}\n")
    
    # Run tests
    runner = TestRunner()
    runner.run_all_tests()
    
    # Print summary and exit with appropriate code
    exit_code = runner.print_summary()
    sys.exit(exit_code)


if __name__ == '__main__':
    main()

